[front] - feat(skills): JIT

JulesBelveze opened the pull request:
## Description

This PR implements Just-In-Time (JIT) skills for conversations. JIT skills are skills that apply to all agents in a conversation (rather than being scoped to a specific agent configuration), allowing users to dynamically enable skills when starting a new conversation.

The implementation makes `agentConfigurationId` nullable in the `conversation_skills` and `agent_message_skills` tables. When `agentConfigurationId` is `NULL`, the skill applies globally to all agents in that conversation.

Key changes:
- Frontend components now pass `selectedSkillIds` through the conversation creation flow
- Added new API endpoint `/api/w/[wId]/assistant/conversations/[cId]/skills` to manage JIT skills
- Created SWR hooks (`useConversationSkills`, `useAddDeleteConversationSkill`) for skill management
- Updated `ConversationResource.upsertSkills` to handle both JIT skills (null `agentConfigurationId`) and agent-scoped skills

## Risk

Database schema migration is required to make `agentConfigurationId` nullable in both `conversation_skills` and `agent_message_skills` tables.

## Tests

Tested locally with the skills feature flag enabled.

## Deploy Plan

1. Run migration 454 to alter the `conversation_skills` and `agent_message_skills` tables
2. Deploy front

JulesBelveze committed:
[front/api] - feature: add API endpoint to manage conversation skills

 - Implement GET handler to fetch all skills associated with a conversation
 - Implement POST handler for adding or deleting a skill to/from a conversation
 - Integrate zod for schema validation of POST request payload
 - Handle errors uniformly and log relevant information for troubleshooting

JulesBelveze committed:
[assistant/conversation] - feature: add ConversationSkillType model to conversation types

 - Introduced ConversationSkillType model to differentiate between custom and global skills within a conversation
 - The new model includes fields for managing skill metadata and timestamps for creation and updates

JulesBelveze committed:
[front/lib/resources] - feature: add fetching and upserting conversation skills

 - Implement `fetchSkills` method to retrieve skills associated with a conversation
 - Implement `upsertSkills` method to update skill associations for a conversation
 - Refactor SQL query in conversation resource for better readability

JulesBelveze committed:
[assistant/conversation] - feature: implement skill addition and deletion in conversations

 - Add useAddDeleteConversationSkill hook to handle skill state changes in conversations
 - Implement addSkill and deleteSkill functionalities in InputBar component
 - Use addSkill and deleteSkill when a skill is selected or deselected, respectively
 - Update ConversationResource error handling for better debugging when an agent configuration is missing
 - Ensure proper propagation of agentConfigurationId for skill management operations

JulesBelveze committed:
[front/lib/models/skill] - refactor: allow nullable agentConfigurationId for JIT skills

 - Modified ConversationSkillModel and AgentMessageSkillModel to accept null agentConfigurationId supporting JIT skills for all agents
 - Created a migration script to alter the database, making agentConfigurationId nullable in related tables

JulesBelveze committed:
[front/types] - feature: extend message and conversation schemas with selectedSkillIds

 - Add `selectedSkillIds` array to track user-selected skills in `MessageBaseSchema`
 - Include `selectedSkillIds` in user message context to enhance message metadata tracking
 - Adjust `ConversationSkillTypeBase` to allow `agentConfigurationId` to be nullable, reflecting optional agent configurations in skill associations

JulesBelveze committed:
[api/w/[wId]/assistant/conversations] - feature: implement JIT skills for conversations

 - Enable adding JIT skills to the conversation upon message posting
 - Make `agent_configuration_id` optional and nullable to support JIT skill flexibility

[api/w/[wId]/assistant/conversations/[cId]/skills] - refactor: update agent configuration parameter

 - Allow `agent_configuration_id` to be optional and nullable in the skill action request schema

JulesBelveze committed:
[front/lib/swr] - feature: implement useConversationSkills hook for skills data fetching

 - Added a new hook to fetch skills data in relation to a conversation including error handling and loading state
 - Integrated a caching mechanism with focus throttle interval of 30 minutes
 - Removed debug console logs from the useAddDeleteConversationSkill function
 - Simplified conditionals by removing redundant checks for agentConfigurationId in skill addition and deletion functions

JulesBelveze committed:
[assistant/conversation] - feature: support selecting skills for conversation messages

 - Add selectedSkillIds to the Conversations API to specify JIT skill on message creation
 - Update InputBar to handle the selection of skills for new conversations
 - Modify ConversationResource for handling skill updates with nullable agentConfigurationId
 - Ensure sync of selectedSkills with conversationSkills in the UI component via hook update

aubin-tchoi commented on front/lib/swr/conversations.ts:
```diff
@@ -428,6 +466,90 @@ export function useAddDeleteConversationTool({
+              action: "add",
+              skill_id: skillId,
+              agent_configuration_id: agentConfigurationId,
+            } as ConversationSkillActionRequest),
```
nit: a `satisfies` would do more work here

aubin-tchoi commented on front/lib/resources/conversation_resource.ts:
```diff
@@ -1173,6 +1176,121 @@ export class ConversationResource extends BaseResource<ConversationModel> {
+    const conversationSkills = await ConversationSkillModel.findAll({
+      where: {
+        workspaceId: auth.getNonNullableWorkspace().id,
+        conversationId: conversation.id,
```
let's add an index for this one

aubin-tchoi commented on front/lib/resources/conversation_resource.ts:
```diff
@@ -1173,6 +1176,121 @@ export class ConversationResource extends BaseResource<ConversationModel> {
+      conversation: ConversationWithoutContentType;
+      skills: SkillResource[];
+      enabled: boolean;
+      agentConfigurationId: string | null | undefined;
```
nit: we probably want only `string | null` and do a `?? null` when calling this

aubin-tchoi commented on front/lib/resources/conversation_resource.ts:
```diff
@@ -1173,6 +1176,121 @@ export class ConversationResource extends BaseResource<ConversationModel> {
+
+      const existingConversationSkill = existingConversationSkills.find(
+        (cs) =>
+          cs.agentConfigurationId === agentConfigurationModelId &&
```
I think `cs.agentConfigurationId` is supposed to be the `sId`, we don't actually need to fetch the agent I think

aubin-tchoi commented on front/lib/models/skill/conversation_skill.ts:
```diff
@@ -81,7 +82,9 @@ export class ConversationSkillModel extends WorkspaceAwareModel<ConversationSkil
-  declare agentConfigurationId: string;
+  declare agentConfigurationId: ForeignKey<
+    AgentConfigurationModel["id"]
+  > | null;
```
should be string | null

aubin-tchoi reviewed (commented):
one main question about `message.context`

JulesBelveze committed:
[front/lib/swr] - refactor: improve error logging for conversation skills

 - Replace console.error with datadogLogger.error to normalize error reporting for skill addition and removal in conversations
 - Include conversationId, workspaceId, and skillId in the logging context for better traceability

JulesBelveze committed:
[front/lib/swr] - feature: implement user notifications for skill add/remove actions

 - Integrates sending user notifications with the addition and removal of conversation skills
 - Enhances error handling by providing user feedback via notifications when skill operations fail

JulesBelveze committed:
[front/lib/resources] - fix: ensure agentConfigurationId is always non-undefined

 - Unified the type of `agentConfigurationId` to be string or null, eliminating the possibility of it being undefined

[front/migrations] - feature: add SQL index for conversation_skills on workspace and conversation IDs

 - Created a new database index to optimize queries filtering by both `workspaceId` and `conversationId` in the `conversation_skills` table

[front/pages/api] - fix: default agentConfigurationId to null when undefined

 - Modified API behavior to explicitly set `agentConfigurationId` to null instead of passing undefined values to maintain consistency

JulesBelveze committed:
[front/lib/models/skill] - refactor: remove AgentConfigurationModel dependency

 - Simplified agentConfigurationId type to a string to decrease coupling
 - Removed unused AgentConfigurationModel import for cleaner codebase

JulesBelveze committed:
[front/hooks] - feature: enable linking skills to new conversations

 - Add support for selecting skills when creating new conversations
 - Include selectedSkillIds in the conversation creation payload

JulesBelveze committed:
[front/lib/resources] - fix: ensure agentConfigurationId is stored as a string

- Cast agentConfigurationModelId to string before assigning to prevent type mismatches
- Handle null values by setting agentConfigurationId to undefined if agentConfigurationModelId is not provided

[front/types/assistant] - fix: update ConversationSkillTypeBase agentConfigurationId type

- Change agentConfigurationId type from ModelId to string to align with updated resource handling

aubin-tchoi commented on front/pages/api/w/[wId]/assistant/conversations/[cId]/skills.ts:
```diff
@@ -0,0 +1,207 @@
+
+  switch (req.method) {
+    case "GET":
+      try {
```
do we need this try catch?

JulesBelveze committed:
[front/lib/resources] - refactor: move conversation skill management to SkillResource

 - Conversation skill fetching and upserting logic has been moved from ConversationResource to SkillResource
 - Removal of irrelevant imports and codes from ConversationResource as part of clean-up

JulesBelveze committed:
fix: lint/format

aubin-tchoi reviewed (commented):
still thinking a bit about the resource part, I think we can have something a bit simpler and more self contained

JulesBelveze committed:
[db] - refactor: update index creation strategy for conversation skills table

 - Changed index creation on "conversation_skills" to use CONCURRENTLY for minimal locking during index building
 - Renamed index to follow naming conventions with "workspace_id" and "conversation_id" in the name

JulesBelveze committed:
[skill_resource] - refactor: streamline skill to conversation upsert logic

 - Simplified the method for updating conversation skills by making `isGlobal` public and modifying the upsert flow
 - Abstracted away direct `ConversationSkillModel` manipulation into the `SkillResource` class

[conversations] - refactor: optimize hook for adding or deleting conversation skills

 - Removed unnecessary memoization for the `useConversationSkills` hook return value
 - Streamlined the payload structure in `useAddDeleteConversationSkill` hook for POST requests

[api] - refactor: clean up GET and POST endpoints for conversation skills

 - Replaced `fetchConversationSkills` with the more accurate `fetchConversationSkillRecords` when getting skills
 - Improved the server-side API endpoint handler for adding and deleting skills to be more concise and readable

[types] - refactor: remove unused `ConversationSkillType` from conversation types

 - Eliminated redundant `ConversationSkillTypeBase` interface following refactors in how conversation skills are handled

JulesBelveze committed:
fix: lint/format

aubin-tchoi reviewed (approved):
LGTM
